x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/PasswordBasedEncryptor.java

```diff
@@ -0,0 +1,232 @@
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Encrypts and decrypts using a password. Decryption authenticates the cyphertext so as to make sure that
+ * the same password has been used during encryption (the cipher mode is AES/GCM/NoPadding). The caller must
+ * ensure that the password and the ciphertext are not stored on the same "medium" (storage partition).
+ * <p>
+ * The {@code password} constructor argument is used to generate AES 256-bit wide keys using the PBKDF2 algorithm.
+ * The "salt", which is the other required parameter to the PBKDF2 algo, is generated randomly (32 byte-wide) using a
+ * {@code SecureRandom} instance. The "salt" is not a secret, like the password is, and it is used to generate different
+ * keys starting from the same password.
+ * <p>
+ * A new encryption key is generated for every {@link PasswordBasedEncryptor} instance (using a newly generated random
+ * "salt"). The key is then reused for as many as {@link #ENCRYPT_INVOKE_LIMIT_USING_SAME_KEY} encryption invocations;
+ * when the limit is exceeded, a new key is computed from a newly generated "salt". In order to support the decryption
+ * operation, the "salt" is prepended to the returned ciphertext. Decryption reads-in the "salt" and uses the secret
+ * password to regenerate the same key and decrypt and authenticate the ciphertext. The key thus computed is locally
+ * cached for possible reuses because generating the key from the password is an expensive operation (by design).
+ * <p>
+ * The reason why there is an encryption invocation limit for the same key is because the AES/GCM/NoPadding encryption mode
+ * must not be used with the same key and the same Initialization Vector. During encryption, the {@link PasswordBasedEncryptor}
+ * randomly generates a new 12-byte wide IV, and so in order to limit the risk of a collision, the key must be changed
+ * after at most {@link #ENCRYPT_INVOKE_LIMIT_USING_SAME_KEY} IVs have been generated and used with that same key. For more
+ * details, see Section 8.2 of https://csrc.nist.gov/publications/detail/sp/800-38d/final .
+ */
+public final class PasswordBasedEncryptor {
+
+    // the count of keys stored so as to avoid re-computation
+    static final int ENCRYPTION_KEY_CACHE_SIZE = 512;
+    // The cipher used to encrypt the data
+    static final String CIPHER_ALGO = "AES";
+    // The mode used with the cipher algorithm
+    private static final String CIPHER_MODE = "GCM";
+    // The padding used with the cipher algorithm
+    private static final String CIPHER_PADDING = "NoPadding";
+    // the KDF algorithm that generate the symmetric key given the password
+    private static final String KDF_ALGO = "PBKDF2WithHmacSHA512";
+    // parameter for the KDF function, it's a funny and unusual iter count larger than 60k
+    private static final int KDF_ITER = 61616;
```

Comment:
Stream of thougths, not necessarily actionable comment: Can you share some insights on how did we decide this? I don't know what can we consider acceptable performance-wise without some benchmarking but given that the keys that PBKDF2 produces are never stored on disk, the attacker we want to protect against is one that tries to brute-force the decryption of a ciphertext ( as opposed to checking the PBKDF2 output for equality in the case of brute-forcing password hashes ). A high enough number would still offer valid protection though, especially as a mitigation factor  for dictionary attacks where users select common passwords. 

Most concrete recommendations I have found for the iteration count, specifically handle the cases of PBKDF2 usage for password hashing/storage, so I guess we should go with as high count as possible without making encryption/decryption too slow.

Reply:
I have chosen the value by looking into various unofficial online sources. It is larger than 60K. It is also larger than what we use for the elasticsearch keystore. It is not  exactly 60K because dictionaries might exist for that exact value.

Even if we don't store the key on disk anywhere, there might be cases where the plaintext is actually known (the snapshot generates metadata-like files that can be known). In this case, the attacker can brute force passwords, by generating keys and then encrypting the known plaintext to check against the captured ciphertext. Because AES encryption is relatively fast, the brute force potential still rests on the speed of generating AES keys, which the iter count modulates.

I am open to suggestions on the value of it, but I wouldn't invest time benchmarking it because this is a generic operation for which countless benchmarks have been made already.

Reply:
Thanks for the feedback. Agreed, I didn't mean to imply that there is no need for protection against brute force . There doesn't even need to be a known plaintext scenario for checking, successful decryption is sufficient proof of a correct key. 

> I am open to suggestions on the value of it, but I wouldn't invest time benchmarking it because this is a generic operation for which countless benchmarks have been made already.

I was challenging that fact as the countless existing benchmarks and suggestions seem to revolve around using PBKDF2 for password storage and it should be up to us to determine how high we can go. However, this was not criticism of the selected value and I don't have a counter-proposal nor do I think we should go build the iteration count bike shed . I have nothing against the 60k

> . It is not exactly 60K because dictionaries might exist for that exact value.

We have the random salts to protect us from that, right ?

I'll mark this as resolved so that we don't have to spend more time on it.

Reply:
> > It is not exactly 60K because dictionaries might exist for that exact value.

> We have the random salts to protect us from that, right ?

That's right. I suppose we could make it a round value? I don't think the "unexpectedness" of the iteration count buys us anything. I'll at least remove the misleading comment.